src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -0,0 +1,482 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/impl/PerfLogImp.h>
+#include <ripple/basics/BasicConfig.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <ripple/json/to_string.h>
+#include <ripple/beast/core/CurrentThreadName.h>
+#include <iostream>
+#include <cstdlib>
+#include <atomic>
+#include <utility>
+#include <string>
+#include <unordered_map>
+#include <cstdint>
+
+namespace ripple {
+namespace perf {
+
+void
+PerfLogImp::Counters::populateRpc(std::vector<char const*> const& labels)
+{
+    for ( auto label : labels)
+    {
+        auto const &elem = rpc_.emplace(label, Rpc());
+        if (elem.second)
+        {
+            elem.first->second.label = label;
+            elem.first->second.labelSs =
+                std::make_unique<Json::StaticString>(
+                    elem.first->second.label.c_str());
+        } else
+        {
+            assert(false);
+        }
+    }
+}
+
+void
+PerfLogImp::Counters::populateJq(
+    std::unordered_map<std::underlying_type_t<JobType>,
+    std::string> const& jobs)
+{
+    for (auto const& job : jobs)
+    {
+        auto elem = jq_.emplace(job.first, Jq());
+        if (elem.second)
+        {
+            elem.first->second.label = job.second;
+            elem.first->second.labelSs =
+                std::make_unique<Json::StaticString>(
+                    elem.first->second.label.c_str());
+        }
+        else
+        {
+            assert(false);
+        }
+    }
+}
+
+Json::Value
+PerfLogImp::Counters::countersJson() const
+{
+    Json::Value rpcobj(Json::objectValue);
+    Rpc totalRpc;
+    for (auto const& proc : rpc_)
+    {
+        Json::Value p(Json::objectValue);
+        p[jss::started] = std::to_string(proc.second.started);
+        totalRpc.started += proc.second.started;
```

Comment:
Hold it.  I'm really confused.  I believe `Rpc::started` is supposed to represent a `time_point`, not a `duration`.  If that's correct then I don't think it makes sense to add a `time_point` to a `time_point` like we're doing here to compute `totalRpc.started`.  What am I missing?

Reply:
@scottschurr Actually, it's a counter as described above. I'm adding comments to describe what totalRpc is.

Reply:
:+1: for comments.  Thanks Mark.